source/fem-dem/fluid_dynamics_vans_matrix_free_operators.cc

```diff
@@ -0,0 +1,499 @@
+          for (unsigned int i = 0; i < dim; ++i)
+            {
+              for (unsigned int k = 0; k < dim; ++k)
+                {
+                  gradient_result[i][i] +=
+                    grad_div_gamma(cell, q) *
+                    (vf_value * gradient[k][k] + value[k] * vf_gradient[k]);
+                }
+            }
+        }
+
+      // SUPG Jacobian
+      for (unsigned int i = 0; i < dim; ++i)
+        {
+          for (unsigned int k = 0; k < dim; ++k)
+            {
+              // Part 1
+              for (unsigned int l = 0; l < dim; ++l)
+                {
+                  // +(ɛ(u·∇)δu + ɛ(δu·∇)u - νɛ∆δu)τ(u·∇)v
+                  gradient_result[i][k] +=
+                    tau * vf_value * previous_values[k] *
+                    (gradient[i][l] * previous_values[l] +
+                     vf_value * previous_gradient[i][l] * value[l] -
+                     this->kinematic_viscosity * hessian_diagonal[i][l]);
+                }
+              // +(ɛ∇δp)τ(u·∇)v
+              gradient_result[i][k] +=
+                tau * vf_value * previous_values[k] * (gradient[dim][i]);
+
+              // +(ɛ∂t δu)τ(u·∇)v
+              if (transient)
+                gradient_result[i][k] += tau * previous_values[k] * vf_value *
+                                         ((*bdf_coefs)[0] * value[i]);
+
+              // Part 2
+              for (unsigned int l = 0; l < dim; ++l)
+                {
+                  // +(ɛ(u·∇)u - νɛ∆u)τ(δu·∇)v
+                  gradient_result[i][k] +=
+                    tau * value[k] * vf_value *
+                    (previous_gradient[i][l] * previous_values[l] -
+                     this->kinematic_viscosity *
+                       previous_hessian_diagonal[i][l]);
+                }
+              // +(ɛ∇p - ɛf)τ(δu·∇)v
+              gradient_result[i][k] +=
+                tau * value[k] * vf_value *
+                (previous_gradient[dim][i] - source_value[i]);
+
+              // +(ɛ∂t u)τ(δu·∇)v
+              if (transient)
+                gradient_result[i][k] += tau * value[k] * vf_value *
+                                         ((*bdf_coefs)[0] * previous_values[i] +
+                                          previous_time_derivatives[i]);
+            }
+        }
+
+      integrator.submit_gradient(gradient_result, q);
+      integrator.submit_value(value_result, q);
+
+      if (this->enable_hessians_jacobian)
+        integrator.submit_hessian(hessian_result, q);
+    }
+
+  if (this->enable_hessians_jacobian)
+    integrator.integrate(EvaluationFlags::values | EvaluationFlags::gradients |
+                         EvaluationFlags::hessians);
+  else
+    integrator.integrate(EvaluationFlags::values | EvaluationFlags::gradients);
+}
+
+/**
+ * The expressions calculated in this cell integral are:
+ * (q, ∇·u) + (v,ɛ∂t u) + (v,ɛ(u·∇)u) - (∇·v,ɛp) + ɛν(∇v,∇u) + ν(v,∇u∇ɛ) -
+ * (v,ɛf) (Weak form), plus two additional terms in the case of SUPG-PSPG
```

Comment:
I would suggest to add here the three terms missing as well:
```suggestion
 * (q, ∇·u) + (v,ɛ∂t u) + (v,ɛ(u·∇)u) - (∇·v,ɛp) + ɛν(∇v,∇u) + ν(v,∇u∇ɛ) -
 * (v,ɛf) (Weak form), plus two additional terms in the case of SUPG-PSPG
```